sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java

```diff
@@ -206,6 +206,122 @@ public Future<Boolean> isHealthy(final Channel channel) {
+        final Promise<String> promise = channel.eventLoop().newPromise();
+
+        if (requestManager == null) {
+            reportIssueUnless(logger, !channel.isActive(), channel, "active with no request manager");
+            return promise.setSuccess("active with no request manager");
+        }
+
+        final Timestamps timestamps = requestManager.snapshotTimestamps();
+        final long currentTime = System.nanoTime();
+
+        if (currentTime - timestamps.lastChannelReadNanoTime() < recentReadWindowInNanos) {
+            // because we recently received data
+            return promise.setSuccess(RntbdConstants.RntbdHealthCheckResults.SuccessValue);
+        }
+
+        // Black hole detection, part 1:
+        // Treat the channel as unhealthy if the gap between the last attempted write and the last successful write
+        // grew beyond acceptable limits, unless a write was attempted recently. This is a sign of a nonresponding write.
+
+        final long writeDelayInNanos =
+            timestamps.lastChannelWriteAttemptNanoTime() - timestamps.lastChannelWriteNanoTime();
+
+        final long writeHangDurationInNanos =
+            currentTime - timestamps.lastChannelWriteAttemptNanoTime();
+
+        if (writeDelayInNanos > this.writeDelayLimitInNanos && writeHangDurationInNanos > writeHangGracePeriodInNanos) {
+
+            final Optional<RntbdContext> rntbdContext = requestManager.rntbdContext();
+            final int pendingRequestCount = requestManager.pendingRequestCount();
+
+            logger.warn("{} health check failed due to nonresponding write: {lastChannelWriteAttemptNanoTime: {}, " +
+                    "lastChannelWriteNanoTime: {}, writeDelayInNanos: {}, writeDelayLimitInNanos: {}, " +
+                    "rntbdContext: {}, pendingRequestCount: {}}",
+                channel, timestamps.lastChannelWriteAttemptNanoTime(), timestamps.lastChannelWriteNanoTime(),
+                writeDelayInNanos, this.writeDelayLimitInNanos, rntbdContext, pendingRequestCount);
+
+            String msg = MessageFormat.format(
+                "{0} health check failed due to nonresponding write: (lastChannelWriteAttemptNanoTime: {1}, " +
+                    "lastChannelWriteNanoTime: {2}, writeDelayInNanos: {3}, writeDelayLimitInNanos: {4}, " +
+                    "rntbdContext: {5}, pendingRequestCount: {6})",
+                channel, timestamps.lastChannelWriteAttemptNanoTime(), timestamps.lastChannelWriteNanoTime(),
+                writeDelayInNanos, this.writeDelayLimitInNanos, rntbdContext, pendingRequestCount
+            );
+
+            return promise.setSuccess(msg);
+        }
+
+        // Black hole detection, part 2:
+        // Treat the connection as unhealthy if the gap between the last successful write and the last successful read
+        // grew beyond acceptable limits, unless a write succeeded recently. This is a sign of a nonresponding read.
```

Comment:
Are the reads and writes the end user workloads? 
If so for write-only/heavy below condition might match right?

Reply:
NO - write means request sent (could be a point read request) - so timespan between last byte of request sent and first byte received - because rntbd like http (at least 1) is strictly request/response protocol that works